### PR TITLE
Module 2 - Oauth Google GitHub

### PR DESCRIPTION
**Pull request by sinpai**
* Added Oauth controller and service
* Added Google oauth 
* Added Github oauth
* Fixed all rubocop errors

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,9 +1,11 @@
...
+
 source 'https://rubygems.org'
 ruby '2.4.1'
 
 git_source(:github) do |repo_name|
-  repo_name = '#{repo_name}/#{repo_name}' unless repo_name.include?('/')
-  'https://github.com/#{repo_name}.git'
+  repo_name = "#{repo_name}/#{repo_name}" unless repo_name.include?('/')
+  "https://github.com/#{repo_name}.git"
 end
```
Comment (AleksSenkou) [inline Gemfile]: @sinpai isn't `bundle install` works good without?
```ruby
git_source(:github) do |repo_name|
  ...
end
```

I mean can we just remove it?)

File: app/controllers/home_page_controller.rb [hunk 2]
```diff
@@ -1,8 +1,9 @@
+# frozen_string_literal: true
+
 class HomePageController < ApplicationController
   before_action :authenticate_user!
 
-  def show
-  end
+  def show; end
```
Comment (AleksSenkou) [inline app/controllers/home_page_controller.rb]: @sinpai let's use
```ruby
def show
end
```
style

File: app/services/users/omniauth.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    user = User.where(provider: provider, uid: uid, email: email).first
+    user ||= User.new(
+      provider: provider,
+      uid: uid,
+      email: email,
+      password: Devise.friendly_token[0, 20]
+    )
+    user.skip_confirmation!
+    user.save
+    user
```
Comment (AleksSenkou) [inline app/services/users/omniauth.rb]: @sinpai currently `#user` method is overloaded with logic, let's split it to several methods

File: bin/rails [hunk 4]
```diff
@@ -1,4 +1,6 @@
 #!/usr/bin/env ruby
+# frozen_string_literal: true
+
```
Comment (AleksSenkou) [inline bin/rails]: @sinpai what do you think about skipping this `frozen_string_literal: true`? I mean you can switch this cop off and remove all of `frozen_string_literal: true` 😉

File: config/environments/production.rb [hunk 5]
```diff
@@ -73,35 +25,16 @@
...
-  # Send deprecation notices to registered listeners.
   config.active_support.deprecation = :notify
-
-  # Use default logging formatter so that PID and timestamp are not suppressed.
   config.log_formatter = ::Logger::Formatter.new
 
-  # Use a different logger for distributed setups.
-  # require 'syslog/logger'
-  # config.logger = ActiveSupport::TaggedLogging.new(Syslog::Logger.new 'app-name')
-
```
Comment (AleksSenkou) [inline config/environments/production.rb:101]: 👍

**Review by AleksSenkou — APPROVED** [review]
LGTM with suggestions

File: Gemfile (hunk 1 above)
Comment (sinpai) [reply Gemfile]: Ok 👍

File: app/controllers/home_page_controller.rb (hunk 2 above)
Comment (sinpai) [reply app/controllers/home_page_controller.rb]: Oh, I gave a power to `rubocop -a` and he went like a tornado through my project

File: bin/rails (hunk 4 above)
Comment (sinpai) [reply bin/rails]: ok 👍  It came from suggested .rubocop.yml from chat. A lot of files was changed by `rubocop -a`. I'll fix it.

Comment (AleksSenkou) [reply bin/rails]: @sinpai `frozen_string_literal: true` makes rails faster, and probably should be used where possible. So, it's your decision here - use or not to use `frozen_string_literal`

File: app/services/users/omniauth.rb [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  end
+
+  private
+
+  delegate :provider, :uid, :info, to: :auth
+  delegate :email, to: :info
+
+  def user
+    user = find_user
+    user ||= create_user
```
Comment (AleksSenkou) [inline app/services/users/omniauth.rb:19]: @sinpai isn't 
```ruby
user = find_user
user ||= create_user
user
```

the same as
```ruby
find_user || create_user
```

?

Comment (AleksSenkou) [reply app/services/users/omniauth.rb:19]: @sinpai we can go even further - memoize `user` variable
```ruby
def user
  @_user ||= find_user || create_user
end
```
in order to have ability to reuse `user` method

File: app/services/users/omniauth.rb [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    user = User.new(
+      provider: provider,
+      uid: uid,
+      email: email,
+      password: Devise.friendly_token[0, 20]
+    )
+    user.skip_confirmation!
+    user.save
+  end
+end
```
Comment (AleksSenkou) [inline app/services/users/omniauth.rb:37]: @sinpai let's add a unit test for `Users::Omniauth`. The test file can be located in `rspec/unit/services/users/omniauth_test.rb`

**Review by AleksSenkou — COMMENTED** [review]
Great changes. Glad to see how everything is going 💪 

I've left several suggestions - let's discuss

**Comment by AleksSenkou** [comment]
@sinpai looks like the PR is merged - let's create separate one for changes I've mentioned. What do you think?

**Comment by sinpai** [comment]
@AleksSenkou I haven't seen comments on this PR, sorry. I will create additional PR for this.

**Comment by AleksSenkou** [comment]
@sinpai let's go with additional trello-card too 😉

**Comment by sinpai** [comment]
new PR is here: https://github.com/sinpai/datarockets-finance-app/pull/8

**Pull request merged**